### PR TITLE
Allenwang2004 proposal on ADMM

### PR DESCRIPTION
**Pull request by Allenwang2004**
In this term project, I want to build a high-performance Consensus ADMM optimization toolkit for distributed convex problems.
Built with C++ core + pybind11 bindings for speed, and exposed via a simple Python API for usability.

### CODE REVIEW

**Comment by Allenwang2004** [comment]
@yungyuc Sorry for editing the proposal several time, I'm ready for the review. Thanks.

File: project/Allenwang2004/README.rst [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	•	GitHub Repository URL: https://github.com/Allenwang2004/consensus-admm
+	•	The library provides:
+	•	Low-level C++ core with fast linear algebra routines, parallel updates, and memory-efficient structures.
+
+
+Problem to Solve & Prospective Users
+
+======================================
+
+Because of the increasing size of datasets and models, many optimization problems in machine learning and statistics are naturally distributed across multiple nodes.
```
Comment (yungyuc) [inline project/Allenwang2004/README.rst:19]: Use 80-char line width.

File: project/Allenwang2004/README.rst [hunk 2]
```diff
@@ -0,0 +1,138 @@
+==============================================
+Consensus ADMM 
+==============================================
+
+Basic Information
+
+=================== 
+
+	•	A high-performance Consensus ADMM optimization toolkit designed for solving distributed convex problems in large-scale machine learning and numerical optimization tasks.
```
Comment (yungyuc) [inline project/Allenwang2004/README.rst:9]: Try to use ascii for plain text.  RST list syntax uses ascii.

File: project/Allenwang2004/README.rst [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+Because of the increasing size of datasets and models, many optimization problems in machine learning and statistics are naturally distributed across multiple nodes. 
+Consensus ADMM is a powerful algorithm to solve such problems efficiently by decomposing them into smaller subproblems that can be solved in parallel while enforcing consensus among the variables.
+
+These problems typically appear in:
+    •	Distributed Lasso / ElasticNet regression: split data across machines, reach shared model.
+    •	Federated Learning: multiple clients optimize local models while maintaining global consistency.
+    •	PCA / Matrix Completion: decompose large matrices with consensus constraints.
+
+I will finish the part of regression model. If I have time, I will implement the part of PCA / Matrix Completion.
```
Comment (yungyuc) [inline project/Allenwang2004/README.rst:27]: PCA (itself, not its applications) could be easier to do, FYI.

File: project/Allenwang2004/README.rst [hunk 4]
```diff
@@ -0,0 +1,138 @@
+==============================================
+Consensus ADMM 
+==============================================
+
+Basic Information
+
+=================== 
+
+	•	A high-performance Consensus ADMM optimization toolkit designed for solving distributed convex problems in large-scale machine learning and numerical optimization tasks.
+	•	GitHub Repository URL: https://github.com/Allenwang2004/consensus-admm
```
Comment (yungyuc) [inline project/Allenwang2004/README.rst:10]: Your project name is a little bit long, but I can tolerate.

It's hard for you to pronounce.

File: project/Allenwang2004/README.rst [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	•	Python: rich docstrings + examples
+
+Schedule
+
+==========
+
+Week Milestone
+
+* 09/27	C++ core skeleton (variable structs, solver loop, regression model)
+* 10/04	Implement primal/dual updates, test with simple quadratic
```
Comment (yungyuc) [inline project/Allenwang2004/README.rst:122]: How do you plan to make automatic tests for C++ within a calendar week without Python wrapper?

If you have not thought about the question, come up with a solution, and change your schedule if needed.

File: project/Allenwang2004/README.rst [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+* 10/11	Python pybind11 wrapper + basic API
+* 10/18	Add support for L1, L2 proximal operators
+* 10/25	Benchmark on synthetic distributed Lasso && Ridge regression
+* 11/01	Add plotting / convergence diagnostics
+* 11/08	Optimize performance, memory usage; add multi-threading
```
Comment (yungyuc) [inline project/Allenwang2004/README.rst:127]: Multi-threading can easily take a work-month, FYI.

File: project/Allenwang2004/README.rst [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+* 11/15	Optimize performance, memory usage; add multi-threading
+* 11/22	Write docs, test edge cases, finalize README
```
Comment (yungyuc) [inline project/Allenwang2004/README.rst:129]: If you can plan for edge cases, they will not be edge cases.  Absorb the handling of edge cases in other bullets.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Multiple commits are OK.  I will ask for squashing if they are too many.  Points to address:

- [ ] Use 80-char line width.
- [ ] Consider the order of automatic C++ testing and Python wrapper?

Review protocol: ask for a next review when you are ready by making a global comment summarizing what you did, along with inline comments highlighting the specific points.

File: project/Allenwang2004/README.rst [hunk 8]
```diff
@@ -83,56 +94,58 @@ Python Interface
...
-* 09/27	C++ core skeleton (variable structs, solver loop, regression model)
-* 10/04	Implement primal/dual updates, test with simple quadratic
-* 10/11	Python pybind11 wrapper + basic API
-* 10/18	Add support for L1, L2 proximal operators
-* 10/25	Benchmark on synthetic distributed Lasso && Ridge regression
-* 11/01	Add plotting / convergence diagnostics
-* 11/08	Optimize performance, memory usage; add multi-threading
-* 11/15	Optimize performance, memory usage; add multi-threading
-* 11/22	Write docs, test edge cases, finalize README
+* 09/27  Repository setup, literature review, project skeleton(C++ core + Python binding)
```
Comment (Allenwang2004) [inline project/Allenwang2004/README.rst:133]: I will C++ to do the implement first, and use pybind11 later

File: project/Allenwang2004/README.rst [hunk 9]
```diff
@@ -83,56 +94,58 @@ Python Interface
...
+* 10/04  Implement primal/dual updates and support for L1, L2 proximal operators
+* 10/11  Benchmark on synthetic distributed Lasso && Ridge regression
+* 10/18  Python binding with Pybind11
+* 10/25  Add plotting / convergence diagnostics
+* 11/01  Extend to PCA / Matrix Completion via ADMM formulation
```
Comment (Allenwang2004) [inline project/Allenwang2004/README.rst:138]: Since I just need to change the target function, I think I can finish it

File: project/Allenwang2004/README.rst [hunk 10]
```diff
@@ -83,56 +94,58 @@ Python Interface
...
+* 11/08  Optimize performance, memory usage; profile bottlenecks
```
Comment (Allenwang2004) [inline project/Allenwang2004/README.rst:139]: Since multithreading take longer time, I will just try to improve the performance

**Review by Allenwang2004 — COMMENTED** [review]
@yungyuc  I am ready for the review, thanks.

**Comment by yungyuc** [comment]
> @yungyuc I add an entry, but it is ok to have the conflict? I'm not sure how to fix that.

Please figure out how to fix the merge conflict.  Why don't you leave a message on the discord channel to seek help?

**Review by yungyuc — CHANGES_REQUESTED** [review]
- [ ] Use relative path to link to your proposal like others do.

Rebase instead of merging from upstream.  Also squash the commits to one so that I do not need to squash for you when merging.

**Comment by Allenwang2004** [comment]
@yungyuc I squashed the commits and rebase it. Thanks.

File: project/Allenwang2004/README.rst [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+Schedule
+==========
+
+Week Milestone
+
+* 09/27  Repository setup, literature review, project skeleton(C++ core + Python binding)
+* 10/04  Implement primal/dual updates and support for L1, L2 proximal operators
+* 10/11  Benchmark on synthetic distributed Lasso && Ridge regression
+* 10/18  Python binding with Pybind11
```
Comment (yungyuc) [inline project/Allenwang2004/README.rst:136]: Creating the binding so late will also delay your testing.  You will code slower.

File: project/Allenwang2004/README.rst (hunk 10 above)
Comment (yungyuc) [reply project/Allenwang2004/README.rst:139]: You can make multi-threading a stretch goal.

File: project/README.rst [hunk 12]
```diff
@@ -72,6 +72,9 @@ Follow the format to add your project:
 7. `Duck997 <https://github.com/Duck997>`__ for
    `VLSIGR <Duck997/README.rst>`__:
    https://github.com/Duck997/VLSIGR
+8. `Allenwang2004 <https://github.com/Allenwang2004>`__ for
+   `consensus-admm <https://github.com/Allenwang2004/consensus-admm/blob/main/README.rst>`__:
```
Comment (yungyuc) [inline project/README.rst:76]: Let me repeat: use relative path like others do.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Let me repeat: use relative path like others do.

**Comment by yungyuc** [comment]
> @yungyuc I'm rewrite my timeline , add a Stretch goal at the end. Is it appropriate?

It is counter-intuitive to give a stretch goal a time, but it is fine.
